tortoise/organizer/organizer.go

```diff
@@ -0,0 +1,122 @@
+package organizer
+
+import (
+	"context"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+// LayersIterator is used to traverse organized layers.
+// For siplicity LayersIterator iterates over all layers unconditionally.
+type LayersIterator func(types.LayerID)
+
+// Opt is for configuring Organizer.
+type Opt func(*Organizer)
+
+// WithLogger configures logger for Organizer.
+func WithLogger(logger log.Log) Opt {
+	return func(o *Organizer) {
+		o.logger = logger
+	}
+}
+
+// WithLastLayer layer configures last layer that tortoise knows about.
+func WithLastLayer(lid types.LayerID) Opt {
+	return func(o *Organizer) {
+		o.submitted = lid
+		o.received = lid
+	}
+}
+
+// WithWindowSize configures size of the window.
+// By default window of the size 1000 is used.
+func WithWindowSize(size int) Opt {
+	return func(o *Organizer) {
+		o.window = make([]types.LayerID, size)
+	}
+}
+
+// New creates Organizer instance.
+func New(opts ...Opt) *Organizer {
+	o := &Organizer{
+		logger: log.NewNop(),
+		window: make([]types.LayerID, 1000), // ~4kb
+	}
+	for _, opt := range opts {
+		opt(o)
+	}
+	return o
+}
+
+// Organizer is a decorator for tortoise that keeps layers in order within the window
+// and sends them to tortoise only when the window has no gaps.
+type Organizer struct {
+	logger log.Log
+
+	// last submitted layer
+	submitted types.LayerID
+	// last received layer that is not submitted
+	received types.LayerID
+	// all layers between submitted and received
+	// window is non-empty only if there are gaps after submitted
+	window []types.LayerID
+}
+
+// Iterate allows to iterate over all ordered layers that are currently within the window including new layer.
+func (o *Organizer) Iterate(ctx context.Context, lid types.LayerID, f LayersIterator) {
+	if !lid.After(o.submitted) {
+		return
+	}
+	if lid.After(o.submitted.Add(uint32(len(o.window)))) {
+		o.logger.WithContext(ctx).With().Error("tortoise organizer received a layer that does not fit in the current window."+
+			" subbmiting pending with gaps",
+			log.Int("window", len(o.window)),
+			log.Named("submitted", o.submitted),
+			log.Named("received", lid),
+		)
+		o.submitPending(ctx, true, f)
+		o.submitted = lid.Sub(1)
+	}
+
+	// at this stage we know that the layer is within the window
+	pos := int(lid.Uint32()) % len(o.window)
```

Comment:
and the scheme using `% len(o.window)` to decide position is clever but not very intuitive. what do you think of using the buffer scheme in mesh to simply make sure trtl apply layers in order? it's not very space efficient but we are talking about small numbers here. see `nextValidLayers`+`maxValidatedLayer` in mesh.go

Reply:
circular buffer is a natural fit for this use case. it is in general very common pattern and it also helps with correctness. i know that the code with the array is robust and will never leak, otherwise, there will be panic. with map the surface for bugs is larger. in general, i don't like to replace it with a map. i will encapsulate add and pop operations to make it more intuitive.

Reply:
i agree that circular buffer is a very common pattern but mostly in libraries (like resource pooling) where performance is critical. my judgement here is that this is more related to application logic and readability/clarity out-weighs performance here.

encapsulating it sounds like a good idea. making it a library so that it can be reused by the valid layers and processed layers in mesh.go is even better.

Reply:
i moved all buffer code to layerBuffer struct, i think the difference with map is minimal wrt to complexity